### ISSUE TITLE
feat(model/v41): implement CED/CSA2 attention with shared KV/index state

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12896]
# feat(model/v41): implement CED/CSA2 attention with shared KV/index state

```routing
lane: model
paths: ["internal/model/v41_attention.go", "internal/model/v41_attention_test.go"]
expected_steps: 7
repo: fak
public_issue: anthony-chaudhary/fak#12640
```

## Parent context
Parent: #12640 (native DeepSeek V4.1 Flash execution and physical qualification).

## Current state
Implementing this leaf's through-line is not yet started; the parent track is open.

## Problem frame
- Centrality: Core ? supplies the attention path the forward assembles.
- P1: advanced - supplies the attention path the forward assembles.
- P2: advanced - supersedes the stale V4-Flash attention framing.
- P3: preserved - generic attention unchanged.
- P4: advanced - fail-closed ratio tests.

## Why now
This leaf is on the critical path for native V4.1 execution and its parent track is open.

## Core through-line
Implement the V4.1 text attention contract: CED/CSA2 with shared-KV and shared-index source
layers, against the pinned reference `inference/model.py` at revision
`dba1be0a40aa45a94ad051997016db3960a90277`. This supersedes the V4 Flash 0731 CSA/HCA framing in
fak#12636 (43-layer / 64x512 geometry); the V4.1 target is 40 layers, width 5120.

## Working spine
Intake -> bounded change -> focused witness -> landed receipt.

## Gold-plating boundary
No KV paging/disk reuse, no GPU kernel, no full forward assembly. Do not implement the stale
V4-Flash-0731 three-branch combiner. Fail closed on any unimplemented ratio rather than falling
through to the generic Q/K/V attention path.

## Witness
```bash
go -C ..\fak test ./internal/model -run 'TestV41Attention' -count=1
```

### File:Line Seam
`internal/model/v41_mhc.go` output feeding the new `v41_attention.go`

### Blast Radius & Affected Lanes
- **Primary Lane:** `model`
- **Unaffected Lanes:** generic attention and V4 Flash 0731 attention paths unchanged

### Quarantined Fallback Mechanism
Unimplemented attention variants return an explicit V4.1 unsupported error; no silent mock and no
generic-attention fallback.

## Scoped Acceptance Criteria
- [ ] [SW-VERIFIED] CED/CSA2 forward matches an independent scalar oracle on a reduced fixture.
- [ ] [SW-VERIFIED] Shared KV/index source-layer resolution is correct.
- [ ] [SW-VERIFIED] Unimplemented ratios fail closed with a typed error.
- [ ] [SW-VERIFIED] Deterministic two-token Prefill/Step test passes.

## Done condition
The scoped acceptance criteria above pass with the cited witness; no sibling lane regresses.

## Acceptance gate
Focused witness passes, package tests are green, vet and 5-gate boundary checks are clean.

## Work estimate
Estimate: 2 points (bounded leaf).

## Overall completion contribution
Contribution: 2/40 points (provisional baseline).

## Completion standard
development

## Closure binding
The resolving commit cites this issue with Fixes/Ref and carries the repository-qualified lane trailer.

**Comment by anthony-chaudhary** [comment]
Scoping witness from a bounded exploration run (2026-09-13). This leaf is the remaining critical-path gap for native V4.1 text execution; the component primitives it depends on are already landed and green.

## Verified current state (public `fak` @ main)

- `internal/model/v41_forward.go:52-58` declares `v41StageCompress` / `v41StageIndexer` as **reserved** stage names, explicitly NOT live: "the reduced text forward does not execute them ... They become live when a leaf lands the compressor/indexer execution." `internal/model/v41_attention.go` does not exist.
- `v41_attention_state.go:1-386` (`V41AttentionState` with `Prefill`/`Step`/`PublishCandidates`/`PublishTopK`/`KVSourceRows`/`IndexKeys`) is landed and exercised by `v41_attention_state_test.go`.
- Sibling component leaves #12895 (mHC), #12897 (Engram), #12898 (router), #12899 (FP8/MXFP4) and #12901 (forward assembly) are all CLOSED; #12900 (independent oracle) and #12908 (session parity) witnesses are green on trunk.

## Exact seam the leaf must land

1. Declare `v41StageCompress` / `v41StageIndexer` as live `v41ForwardStage` constants and gate them via `v41StageErr` so any missing compressor/indexer tensor fails closed (never falls through to the generic Q/K/V attention path in `v41Layer`, `v41_forward.go:377`).
2. Add `internal/model/v41_attention.go`: the CED/CSA2 contraction over the existing `V41AttentionState` shared-KV / shared-index-source slots, pinned to `deepseek-ai/DeepSeek-V4.1-Flash@dba1be0a40aa45a94ad051997016db3960a90277` `inference/model.py:429-790`.
3. Wire the two stages into `forwardV41` (`v41_forward.go:304`) between `v41StageAttention` and `v41StageMoE`.

## Why it was not taken as a bounded leaf this run

`expected_steps: 7` exceeds the orchestrator lane's <=3-step bound, and the `model` lane is currently held by a live lease (`dos arbitrate --lane model` => refuse). Recorded so the next dispatch is seam-anchored and can be split if needed.

Do not claim any native generation or physical parity here; loader -> `Session.Prefill/Step` integration parity remains a separate promotion gate.

**Comment by anthony-chaudhary** [comment]
﻿## Leaf landed: shared-KV source fail-closed (bounded; not full CED/CSA2 execution)

The reduced V4.1 text assembly projects its own per-layer `attn.wkv.weight` and never consumes a KV source layer's shared key/value state. `v41ForwardAdmitted` gated the Engram (#13007) and compressor/indexer (#13006) stages but **never checked `KVSourceLayerIDs`** — it was retained and validated at config parse, yet the forward admission ignored it. A config declaring a shared-KV source inside the decoder stack therefore ran the reduced per-layer-KV forward and emitted logits as if the shared source were absent. The published 40-layer artifact declares `kv_source_layer_ids {2,8,14,20}`, so the hole was live for the real checkpoint.

Landed `f65d7836bdc67266e24c74b84f52dd9cc0d61f60` on `origin/main`:

- `internal/model/v41_forward.go`: new `v41KVSourceForwardAdmitted`, refusing with a typed error wrapping `ErrV41ForwardStage` naming the layer when a declared KV source lies in `[0,NumLayers)`; wired into `v41ForwardAdmitted`. Out-of-range declarations stay admitted, keeping the reduced oracle fixture (`NumLayers=1`, kv sources >= 2) runnable.
- `internal/model/v41_forward_test.go`: `TestV41ForwardKVSourceDeclaredFailsClosed` (RED -> GREEN).

Witness (SW-VERIFIED):
```
go test ./internal/model -run TestV41ForwardKVSourceDeclaredFailsClosed -count=1
  RED   (guard reverted): in-range KV source admission error = <nil> => FAIL
  GREEN (guard restored): PASS
go test ./internal/model/... -count=1                       => ok 146.446s
go vet ./internal/model                                     => exit 0
gofmt -l internal/model/v41_forward.go ..._test.go          => clean
fak-boundary check <files> --fak-dir <worktree>             => PASS (0 violations, 5 gates)
```

This closes the fail-closed admission gap only; it does **not** implement shared KV/index-state execution (#12896's remaining integration work) and does not claim native V4.1 generation or physical parity. `Ref #12640`.

**Comment by anthony-chaudhary** [comment]
Handoff from fak#13006 (landed `origin/main` `ff690b819`).

The CED/CSA2 compressor and lightning-indexer stages now EXECUTE (`internal/model/v41_forward.go`: `v41CompressedRows` + `v41IndexRows`; primitives ported into package `model` in `internal/model/v41_compress_index.go`). The in-range compressor/index fence is lifted.

What remains for THIS leaf: the reduced assembly's per-position sink contraction still reads the projected per-position KV rows; the compressor's pooled latent is currently expanded back onto its causal positions (`kvRows[i] = compressed[i/ratio]`) so the pooling is observable without rewriting the attention path. Your leaf owns mapping the contraction onto the reference's compressed/shared KV cache: consume the compressed key/value stream directly in `V41SparseAttentionSink` (and the shared-source reuse via `V41IndexerPublication.Reuse`), rather than the positional expansion.

The published 40-layer config still refuses at `v41KVSourceForwardAdmitted` (kv sources {2,8,14,20}) and `v41CandidateSourceForwardAdmitted` (source 20); lifting those is gated on this leaf's KV/index-state execution. After #12896/#12890, #13009 supplies the full geometry, then the first physical strix3 bring-up run.

**Comment by anthony-chaudhary** [comment]
Landed and witnessed on trunk (recovery run).

Commit: `079b1d48a` -> merge `133380d90` (pushed to origin/main as `b6a9d2179..133380d90`)
Lane: (fak model)

[SW-VERIFIED] on the landed tree:
- `go -C ..\fak test ./internal/model -run TestV41Attention -count=1` -> ok
- `go -C ..\fak test ./internal/model -count=1` -> ok (240.083s)
- `go -C ..\fak vet ./internal/model` -> exit 0
- `go run ./cmd/fak-boundary check --staged` -> PASS (0 violations, 5 gates)

Acceptance criteria met: CED/CSA2 compressed contraction matches the scalar oracle on a reduced fixture; shared KV/index source-layer resolution correct; unimplemented ratios fail closed with a typed error (no generic Q/K/V fallthrough); deterministic two-token Prefill/Step test passes.

Forward-only provenance note: the landed commit audits `UNATTESTED` (code+tests, no trailers). Trailers were minted after landing and cannot be applied without rewriting published history (a peer commit sits on top). Recorded forward-only in `docs/tickets/deepseek-native/LEAF-2026-09-15-v41-12896-ced-csa2-shared-kv-landed.md` per the halo-agent-suite recovery precedent; that note does not claim an audit repair.

[HW-WITNESSED] physical V4.1 attention execution on strix1-3 remains open under parent epic #12640.
